test/test_meta.py

```diff
@@ -1,657 +1,107 @@
-# this test file (since I could have just inlined __torch_function__ on the
-# OpInfo call, and OpInfos generally have very regular inputs), but it will be
-# useful for more comprehensive testing e.g., as seen in
-# https://github.com/pytorch/pytorch/pull/75994
-class MetaCrossRefMode(torch.overrides.TorchFunctionMode):
-    test_case: TestCase
-    run_excludes_anyway: bool
-
-    def __init__(self, test_case, *, run_excludes_anyway):
-        self.test_case = test_case
-        self.run_excludes_anyway = run_excludes_anyway
+def safe_is_leaf(t):
+    try:
+        return t.is_leaf
+    except RuntimeError:
+        # inference mode can trigger this
+        return False
 
-    def __torch_function__(self, func, types, args=(), kwargs=None):
-        kwargs = kwargs or {}
 
-        hit = 0
-        miss = 0
+# This is a class for converting multiple tensors into meta tensors which
+# share the same view/storage structure.  The operation model is you allocate
+# one of these, and then call it repeatedly on all the tensors you want to
+# convert.  It's important to use the same object for tensors you want to
+# share storage because this is how we correlate shared storages to the same
+# meta storages; similarly, it's important NOT to use the same object for
+# unrelated groups of tensors because this class will remember all the
+# tensors/storages its seen and therefore leak memory.
+class MetaConverter:
+    def __init__(self):
+        self.storage_memo = {}
+        self.tensor_memo = {}
+        self.hit = 0
+        self.miss = 0
 
-        # Doesn't actually return a storage
-        @functools.lru_cache(None)
-        def meta_storage(s):
-            return torch.empty(s.size(), dtype=s.dtype, device='meta')
+    def successful(self):
+        return self.hit > 0 and self.miss == 0
 
-        def safe_is_leaf(t):
-            try:
-                return t.is_leaf
-            except RuntimeError:
-                # inference mode can trigger this
-                return False
+    # NB: doesn't actually return a storage, because meta storage is
+    # not supported
+    def meta_storage(self, s):
+        if s not in self.storage_memo:
+            self.storage_memo[s] = torch.empty(s.size(), dtype=s.dtype, device='meta')
+        return self.storage_memo[s]
 
-        @functools.lru_cache(None)
-        def meta_tensor(t):
+    # This function assumes that it's possible to do the conversion
+    def meta_tensor(self, t):
+        if t not in self.tensor_memo:
             with torch.inference_mode(t.is_inference()):
-                s = meta_storage(t.storage())
+                s = self.meta_storage(t.storage())
                 is_leaf = safe_is_leaf(t)
                 if is_leaf or not t._is_view():
                     r = torch.empty(
                         (0,), dtype=t.dtype, device='meta'
                     )
-                    r.set_(s, t.storage_offset(), t.size(), t.stride())
+                    with no_dispatch():
```

Comment:
just for my own edification - why is this necessary ?

Reply:
It's very dumb... set_ with Tensor argument redispatches to set_ with Storage which would call back into Python when there's a mode... but we can't because meta storages don't work in Python lol